### PR TITLE
Support save/load tf checkpoint to remote and file exists and mkdirs operations for remote

### PR DESCRIPTION
**Pull request by jenniew**
Save/load tf checkpoint to remote without using tensorflow native method.
Support exists and mkdirs operations for remote files.

### CODE REVIEW

File: pyzoo/zoo/orca/data/file.py [hunk 1]
```diff
@@ -92,50 +94,19 @@ def load_numpy(path):
...
-        except Exception as ex:
-            if ex.response['Error']['Code'] == 'NoSuchKey':
-                return False
-            raise ex
-        return True
-    else:
-        return os.path.exists(path)
+    # local path
+    if not (path.startswith("hdfs://") or path.startswith("s3://") or
+                path.startswith("file://")):
```
Comment (hkvision) [inline pyzoo/zoo/orca/data/file.py:99]: local file path can also start with file://?

File: pyzoo/zoo/orca/data/file.py [hunk 2]
```diff
@@ -92,50 +94,19 @@ def load_numpy(path):
...
-        key = "/".join(path_parts)
-        try:
-            s3_client.get_object(Bucket=bucket, Key=key)
-        except Exception as ex:
-            if ex.response['Error']['Code'] == 'NoSuchKey':
-                return False
-            raise ex
-        return True
-    else:
-        return os.path.exists(path)
```
Comment (hkvision) [inline pyzoo/zoo/orca/data/file.py:117]: For local file we can just call this?

File: pyzoo/zoo/orca/learn/tf/utils.py [hunk 3]
```diff
@@ -167,3 +174,107 @@ def combine(pair):
...
+    temp = tempfile.mkdtemp()
+    if saver is None:
+        saver = tf.train.Saver()
+    saver.save(sess, join(temp, ckpt_name))
+    # change checkpoint file
+    with open(join(temp, "checkpoint")) as f:
+        new_lines = []
+        lines = f.readlines()
+        # replace model_checkpoint_path and all_model_checkpoint_paths to checkpoint name
+        #  instead of absolute checkpoint path
```
Comment (hkvision) [inline pyzoo/zoo/orca/learn/tf/utils.py:198]: # instead of the absolute checkpoint path

File: pyzoo/zoo/orca/learn/tf/utils.py [hunk 4]
```diff
@@ -167,3 +174,107 @@ def combine(pair):
...
+        for line in lines:
+            if re.compile("^model_checkpoint_path: \"(.*)\"$").match(line):
+                new_lines.append("model_checkpoint_path: \"{}\"\n".format(ckpt_name))
+            elif  re.compile("^all_model_checkpoint_paths: \"(.*)\"$").match(line):
```
Comment (hkvision) [inline pyzoo/zoo/orca/learn/tf/utils.py:202]: remove the redundant space before re

File: pyzoo/test/zoo/orca/data/test_file.py [hunk 5]
```diff
@@ -68,15 +68,6 @@ def test_exists_local(self):
         file_path = os.path.join(self.resource_path, "orca/data/abc.npy")
         assert not exists(file_path)
 
-    def test_exists_s3(self):
```
Comment (hkvision) [inline pyzoo/test/zoo/orca/data/test_file.py:71]: so is S3 supported for now?

File: pyzoo/zoo/orca/learn/tf/utils.py [hunk 6]
```diff
@@ -167,3 +174,107 @@ def combine(pair):
...
+    with open(join(temp, "checkpoint")) as f:
+        lines = f.readlines()
+        # get checkpoint name from 'checkpoint' file
+        for line in lines:
+            m = re.compile("^model_checkpoint_path: \"(.*)\"$").match(line)
+            if m:
+                ckpt_name = m.group(1)
+                break
+    # filter checkpoint files
+    checkpoint_files = [file for file in file_list if basename(file).startswith(ckpt_name)]
```
Comment (hkvision) [inline pyzoo/zoo/orca/learn/tf/utils.py:240]: error handling if ckpt_name is not found?

**Comment by hkvision** [comment]
LGTM after the above comments are addressed.

File: pyzoo/zoo/orca/data/file.py (hunk 1 above)
Comment (jenniew) [reply pyzoo/zoo/orca/data/file.py:99]: yes, I mean the file path does not have prefix. I may change the comment.

File: pyzoo/test/zoo/orca/data/test_file.py (hunk 5 above)
Comment (jenniew) [reply pyzoo/test/zoo/orca/data/test_file.py]: s3 change back to use botos. may change later to solve extra jar problem.

**Comment by jenniew** [comment]
Run Jenkins. It only fails for JEP test which is not related to this PR.

**Pull request merged**